lib/libzrepl/data_conn.c

```diff
@@ -749,7 +765,10 @@ uzfs_zvol_rebuild_dw_replica(void *arg)
 				}
 				LOG_INFO("Rebuild started from clone for vol "
 				    "%s", zinfo->name);
+
+				wait_snap_timeout();
```

Comment:
should we do the wait before changing state to AFS? Becz, in AFS, we write application incoming IOs to both main_zv, clone_zv, and, if helping replica breaks the connection bit late, there is chance of incoming IOs going to main_zv and missing snapshot can't be taken after that.

Reply:
we have to wait after changing the state to AFS only because if we are waiting before changing the state then there is again a problem of getting a snapshot command before degraded replica is in AFS mode in which case it will ignored.

Also helping replica will break the connection if there is any snapshot command in the pipeline, and if there is any snapshot command then there won't be any IOs. so there is no chance of incoming IOs going to main_zv. Let me know if it answers your question.

Reply:
there can be incoming IOs after target waits for SNAP_TIMEOUT. But, if helping replica takes more than SNAP_TIMEOUT seconds to break the rebuild connection, degraded replica will start getting IOs, and it will be written to both main_zv and clone_zv.

Reply:
New thoughts:
how about a state AFS_IN_TRANSITION?
Before changing to AFS, let degraded replica move it to AFS_IN_TRANSITION. In AFS_IN_TRANSITION also, degraded will break rebuild if snapshot command is received, but, incoming IOs will not be written to main_zv.
Let helping replica wait for 2*SNAP_TIMEOUT, before it starts sending data to degraded replica.
Degraded replica will be in AFS_IN_TRANSITION until it gets some response from helping replica, and then, changes to AFS state, and also, rebuild from self clone_snapshot.
So, let helping replica send a message to degraded replica after it comes out of sleep.

Reply:
@satbirchhikara ^^^